FIX 'which' check in 'sake' now works for aliases

NoLdman opened the pull request:
## Description

`sake` may wrongly exit early on its `which` check, since the current implementation does the following:

1. Retrieves via `command -v` information about the `which` executable
2. Tries to check the returned information string, if it is executable.

This solution is problematic since the returned information string may include alias information (and more),
which `test -x` can not interpret.

Additionally is the `test -x` check unnecessary  since `command -v` already only returns executable command or fails if the command can not be found.

Therefore, I changed the source to just check for the exit status of `command -v`.

## Manual testing steps

Run `vendor/bin/sake dev/build` on a POSIX compliant system, where `which` itself has been aliased.

## Issues

- https://github.com/silverstripe/silverstripe-framework/issues/10682


## Pull request checklist

- [x] The target branch is correct
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

NoLdman committed:
FIX 'which' check in 'sake' now works for aliases

Improves on the current implementations since
'command' already only returns executable commands.
Additionally, the success output of 'command' is not standardized
and may contain additional information.
Therefore, a 'test' if the result of 'command'
is a valid executable file, may fail incorrectly.

Now just relying on the standardized exit status of 'command -v'.

GuySartorelli commented on sake:
```diff
@@ -9,10 +9,7 @@ Executes a SilverStripe command"
-  echo "Error: sake requires the 'which' command to operate." >&2
-  exit 1
-fi
+command -v which >/dev/null 2>&1 || { echo >&2 "Error: sake requires the 'which' command to operate."; exit 1; }
```
```suggestion
command -v which >/dev/null 2>&1
if [ $? -ne 0 ]; then
  echo "Error: sake requires the 'which' command to operate." >&2
  exit 1
fi
```
I'd prefer to keep the actual condition in a more traditional conditional statement for easy readability.

minimalic replied:
Yes, it works. I've tested both ways on macOS, Rocky Linux/RHEL and Ubuntu (`bash` environment).

michalkleiner replied:
looks sensible to me with your tweak for better legibility

GuySartorelli reviewed (changes requested):
[Docs for the `command` command](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/command.html) for future reference.

@michalkleiner and @minimalic can you please confirm if this works for you? I note there was discussion about the [alternative PR](https://github.com/silverstripe/silverstripe-framework/pull/11218), but hopefully this PR works for all those cases.

NoLdman committed:
ENH Improved readability of sake's which check

The previous implementation used '&&' piping
wich is handled differently in PHP and Bash code
and therefore, may be confusing to a PHP developer.
I refactored the function to use a traditional if statement.

GuySartorelli commented on sake:
```diff
@@ -9,7 +9,7 @@ Executes a SilverStripe command"
 fi
 
-if ! [ -x "$(command -v which)" ]; then
+if ! command -v which >/dev/null 2>&1; then
```
```suggestion
command -v which >/dev/null 2>&1
if [ $? -ne 0 ]; then
```
Unless there are reasons against doing it this way, please do it like this per my previous request. 

The test syntax is consistent with other conditions in this file and is an expected syntax for conditions.
Explicitly testing the exit code makes it obvious that that's all we actually are interested in for this test.

NoLdman replied:
Wow, that was a fast response! :hushed: - I was typing my reasoning, but you already answered. 

I interpreted your problem as follow:

1. PHP and Bash  handle '&&' piping differently and therefore may be confusing to PHP developers
2. The fact that the line exceeded 100 characters.

The reason behind my adaptation to your suggestion was the following:

- `$? -ne 0` is some syntax I only know of Bash scripting and therefore may also be very unfamiliar to a PHP developer.
- With the combination of the two lines I moved them clearer into one context: "If some command doesn't work, then…" (in contrast to your code of: "If a magic variable has something to do with "0"…)

If you don't agree with that reasoning, I'm fine to use your suggestion instead.

GuySartorelli replied:
I think it's totally fine to use bash constructs in a bash context ;p

But I also think that testing a specific thing makes it easier to look up what that thing is. I.e, testing `$? -ne 0` makes it easy to look up what that specific syntax does (i.e. it's testing if the exit code of the previously executed command is not equal to 0).

If the full condition is `! command -v which >/dev/null 2>&1` then you need to know what *all* of that does and will return in order to know that what is being tested is "If some command doesn't work", and not "if this thing returns something falsy" like you'd expect if you're more familiar with php conditions

For that reason I'd prefer the separated syntax in my suggestion

GuySartorelli replied:
Basically, at some point someone with limited bash experience will see this and be confused regardless of which syntax we use - so by breaking it up into small chunks (separsting out the command from the test) it should be easier to find out what's happening, imo

NoLdman replied:
Yeah, it's a difficult topic since everyone has different experiences, know-how, and approaches coding differently.

Therefore, on such philosophical topics, I'm always siding with the team of that particular project.

minimalic replied:
We're talking about a line of code that checks the availability of `which` for days now.

I mean how many users will execute `sake` without having `which` installed? Zero? :slightly_smiling_face:


BTW I've only used the `type -P` approach to let the code be as readable as it was before, even for non-bash developers:
```sh
if ! [ -x "$(type -P which)" ]; then
  echo "Error: sake requires the 'which' command to operate." >&2
  exit 1
fi
```
The `command -v` is also my personal preferred way. 👍 The @GuySartorelli's code looks easier to me while @NoLdman's code is more elegant.

NoLdman committed:
ENH Improved readability of sake's which check

Converted to 'if' to square bracket based testing
to be in comply with the other 'if' statements in the script.

Co-authored-by: Guy Sartorelli <36352093+GuySartorelli@users.noreply.github.com>

GuySartorelli reviewed (approved):
LGTM, works well locally.
Retargetted to `5.2` so it can be released as a patch. Will merge when CI goes green.
